docs/ebnf.md

```diff
@@ -0,0 +1,115 @@
+  (\* interface has one or both but never neither \*)
+
+**function interface** =
+  "interface" , **identifier** , "{" , **function signatures** , "}" ;
+
+**function signature** =
+  [ **identifier** ] , "(" , [ **parameters** ] ,  ")" ,
+  "=>", [ **identifier** , [ "?" ] ":" ] , **type** ,
+  { [ "," ] , **function trailer** } ;
+  (\* parser should confirm semantics, allow 1 of each function trailer \*)
+
+
+## Non-Terminal Production Rules
+
+
+### Repeaters (length >= 1)
+
+**object properties** =
+  **object property** , { "," , **object property** } ;
+
+**function signatures** =
+  **function signature** , { "," , **function signature** } ;
+
+**parameters** = **spread parameter**
+               | **parameter** , { "," , **parameter** } , [ "," , **spread parameter** ]
+
+Note: inspired by [ES2015 FormalParameterList](http://www.ecma-international.org/ecma-262/6.0/index.html#sec-function-definitions)
+
+**identifier** =
+  ( **identifier character** , { **identifier character** } ) - **reserved word** ;
+
+**value expressions** = **value expression** , { "," , **value expression** } ;
+
+
+### Objects
+
+**destructured object** = "{" , **object properties** , "}" ;
+
+**object property** = **parameter**
+                       | **object spread property** ;
+
+**object spread property** = "..." , **identifier** , "?" ;
+
+
+### Parameters
+
+**parameter** = **type**
+              | **named parameter**
+              | **destructured object** ;
+
+**named parameter** =
+  **identifier** ,
+  ( [ **default value** ] ":" , **type** )
+  | ( **default value** ":" , [ **type** ] );
+  (\* named parameter has one or both but never neither \*)
+
+**default value** = "=" , **literal type** ;
+
+**spread parameter** = "..." , **identifier** , ":" , **array type** ;
+
+**function trailer** = ( "requires" , ":" , **value expressions** )
+                     | ( "throws" , [ ":" , **identifier** ] ) ;
+
+
+### Types
+
+**type** = **array type**
+         | **non-array type**
+         | **union type** ;
+
+**union type** = **type** , { "|" , **type** } ;
+
+**array type** = **non-array type** , "[" , "]" ;
+
+**non-array type** = **identifier**
+                   | **builtin type**
+                   | **literal type** ;
+
+
+## Terminal Production Rules and JavaScript
+
+**literal type** = ? set of JavaScript-legal literal values ? | **regexp literal**;
+
+**predicate literal** = ? JavaScript-legal arrow function ? , ";" ;
+
+**regexp literal** = "/" , ? JavaScript-legal Regular Expression characters ? , "/" ;
+
+**identifier character** = ? set of JavaScript-legal identifier characters ? ;
+
+**reserved word** = ? set of JavaScript and rtype reserved words ? ;
+
+**builtin type** = "Any"
+                 | "Array"
+                 | "Boolean"
+                 | "Function"
+                 | "Number"
+                 | "Object"
+                 | "String"
+                 | "Void"
+                 | "Predicate" ;
```

Comment:
ECMAScript specifies lots of global constructors:
- http://www.ecma-international.org/ecma-262/6.0/index.html#sec-constructor-properties-of-the-global-object

Should we be parsing these out explicitly? Or is there a way to refer to an existing constructor in the general sense to keep the EBNF simple?